File: 2024-06-June/6-03-24_2486.c
/* 
 * 2486. Append Characters to 
 *       String to Make Subsequence
 *
 *  You are given two strings s and t,
 *  consisting of only lowercase English letters
 *
 *  Return the minimum number of characters
 *  that need to be appended to the end of s
 *  so that t becomes a subsequence of s.
 *
 *  A subsequence is a string that can be
 *  derived from another string by
 *  deleting some or no characters
 *  without changing the order of the 
 *  remaining characters.
 *
 *  Decided to write this in C 
 *  since the constraints to this
 *  problem were very slim, and 
 *  that the answer was ez.
 *  O(n) time, O(1) space
 * */
#include <string.h>

int appendCharacters(char* s, char* t) {
    unsigned int sp = 0;
    unsigned int tp = 0;
    unsigned int sn = strlen(s);
    unsigned int tn = strlen(t);

    while(sp < sn && tp < tn){
        if(s[sp] == t[tp]){
            sp++;
            tp++;
        } else sp++;
    }
    return tn - tp;
}

File: 2024-06-June/6-03-24_2486.py
'''
2486. Append Characters to String
        to Make Subsequence

Difficulty: Medium
Topics: Two Pointers, String, Greedy

You are given two strings s and t 
consisting of only lowercase English letters.

Return the minimum number of characters 
that need to be appended to the end of s 
so that t becomes a subsequence of s.

A subsequence is a string that can be derived 
from another string by deleting some or no 
characters without changing 
the order of the remaining characters.

This problem is extremely greedy.
The "medium" tag really threw me
off since the answer is easy.
O(n) time, O(1) space
'''

class Solution:
    def appendCharacters(self, s: str, t: str) -> int:
        sp, tp = 0, 0
        sn, tn = len(s), len(t)
        while sp < sn and tp < tn:
            if s[sp] == t[tp]:
                sp += 1
                tp += 1
            else:
                sp += 1
        
        return tn - tp
        
        

